# Importing the necessary libraries

In [1]:
import pandas as pd
import numpy as np

# Loading the data

In [2]:
fuel_data = pd.read_csv('fuel_ferc1.csv')

# Question 1

In [3]:
A = [1, 2, 3, 4, 5, 6]
B = [13, 21, 34]

In [4]:
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

In [5]:
C = [1, 2, 3, 4, 5, 6]
D = [13, 21, 34]

In [6]:
C.append(D)
C

[1, 2, 3, 4, 5, 6, [13, 21, 34]]

#### A.extend(B) is the right answer

# Question 2

In [7]:
print(np.identity(3))

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [8]:
#eye(3) gives NameError: name 'eye' is not defined
#np.array([1,0,0],[0,1,0],[0,0,1]) gives ValueError: only 2 non-keyword arguments accepted
#np.array[(1,0,0),(0,1,0),(0,0,1)] gives TypeError: 'builtin_function_or_method' object is not subscriptable

#### np.identity(3) is the right answer

# Question 3

In [9]:
fuel_data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [10]:
list(fuel_data)

['record_id',
 'utility_id_ferc1',
 'report_year',
 'plant_name_ferc1',
 'fuel_type_code_pudl',
 'fuel_unit',
 'fuel_qty_burned',
 'fuel_mmbtu_per_unit',
 'fuel_cost_per_unit_burned',
 'fuel_cost_per_unit_delivered',
 'fuel_cost_per_mmbtu']

In [11]:
a = fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].sum()
b = fuel_data['fuel_type_code_pudl'].value_counts()

a.div(b)

coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
dtype: float64

#### Gas is the fuel type with the lowest average fuel cost per unit burned.

# Question 4

In [12]:
print(np.percentile(fuel_data['fuel_mmbtu_per_unit'], 75))
print(np.std(fuel_data['fuel_mmbtu_per_unit']))

17.006
10.600040781504145


#### 10.60 and 17.01 respectively

# Question 5

In [13]:
print(pd.DataFrame(fuel_data['fuel_qty_burned']).skew()[0])
print(pd.DataFrame(fuel_data['fuel_qty_burned']).kurtosis()[0])

15.851495469109503
651.3694501337732


#### 15.85 and 651.37 respectively

# Question 6

In [14]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [15]:
c = fuel_data.isnull().sum()[5]
d = fuel_data.shape[0]

print('percentage of missing rows = {}%'.format((c/d)*100))

percentage of missing rows = 0.609694136774718%


#### Feature with missing values: fuel_unit, Total: 180, Percentage to 3dp: 0.610%

# Question 7

In [16]:
fuel_data['fuel_unit'].value_counts()

mcf       11354
ton        8958
bbl        7998
gramsU      464
mmbtu       180
kgU         110
mwhth       100
mwdth        95
gal          84
Name: fuel_unit, dtype: int64

#### This implies that the missing values are in categories (are not continuous) and we replace or fill them with the mode.

# Question 8

In [17]:
fuel_data.corr()['fuel_cost_per_unit_burned'].groupby(level = 0).max()

fuel_cost_per_mmbtu            -0.000437
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_mmbtu_per_unit            -0.010034
fuel_qty_burned                -0.018535
report_year                     0.013599
utility_id_ferc1               -0.037863
Name: fuel_cost_per_unit_burned, dtype: float64

#### fuel_qty_burned and fuel_mmbtu_per_unit have the second and third lowest correlation respectively.

# Question 9

In [18]:
e = fuel_data[(fuel_data['fuel_type_code_pudl'] == 'coal') & (fuel_data['report_year'] == 1998)]['fuel_cost_per_unit_burned'].sum()
f = fuel_data[(fuel_data['fuel_type_code_pudl'] == 'coal') & (fuel_data['report_year'] == 1994)]['fuel_cost_per_unit_burned'].sum()

print('The percentage change = {}%'.format((e - f)/f * 100))

The percentage change = -20.56765451826053%


#### The percentage change is approximately -21%

# Question 10

In [19]:
g = fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].sum()
h = fuel_data['report_year'].value_counts()

g.div(h)

1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
dtype: float64

#### The year 1997 has the highest average fuel cost per unit delivered